Add/update ceph csi snapshotter rbac and sidecar deployments

humblec opened the pull request:
<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test full]
[test ceph]

Madhu-1 commented on cluster/charts/rook-ceph/templates/clusterrole.yaml:
```diff
@@ -296,6 +296,24 @@ rules:
   - apiGroups: [""]
     resources: ["persistentvolumeclaims/status"]
     verbs: ["update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
```
need same changes in common.yaml and e2e manifests also

humblec replied:
was already on it, common is already addressed. Looking into e2e.

humblec replied:
done. thanks .

Madhu-1 commented:
added skip CI  to avoid running E2E till we get ceph-csi release out

humblec commented:
> a couple of things needed.
> 
>     * documentation for cephfs snapshot, restore and clone
> 
>     * csi  image version change
> 
> 
> This PR should not be merged till v.3.1.0 is released

Yeah.. :+1:

humblec commented:
I will push a documentation PR as a followup.

Madhu-1 commented on cluster/examples/kubernetes/ceph/common.yaml:
```diff
@@ -1582,6 +1582,24 @@ rules:
   - apiGroups: [""]
     resources: ["events"]
     verbs: ["list", "watch", "create", "update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
```
These new rbac.need to be added for upgrade yaml file

humblec replied:
Done. added to `1.3` upgrade yaml. Sure, after 1.4.0 they could apply the same and get it working.

humblec replied:
@travisn @Madhu-1 ptal.

humblec commented on cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:
```diff
@@ -34,6 +34,21 @@ spec:
           volumeMounts:
             - name: socket-dir
               mountPath: /csi
+        - name: csi-snapshotter
```
Yeah. that would be good and can both ( RBD and CephFS ) handled together in a followup PR :+1:

Madhu-1 replied:
is it possible to do it before the release?

humblec replied:
we can defer it from this release and get it done for next patch release.

humblec commented:
@travisn @Madhu-1  I have added the documentation too in this PR itself. PTAL .

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
```diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
+
+In [snapshotClass](https://github.com/rook/rook/tree/{{ branchName }}/cluster/examples/kubernetes/ceph/csi/cephfs/snapshotclass.yaml),
+the `csi.storage.k8s.io/snapshotter-secret-name` parameter should reference the
+name of the secret created for the cephfsplugin and `pool` to reflect the Ceph pool name.
```
why we need pool in cephfs snapshot class?

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
```diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
+the `csi.storage.k8s.io/snapshotter-secret-name` parameter should reference the
+name of the secret created for the cephfsplugin and `pool` to reflect the Ceph pool name.
+
+Update the value of the `clusterID` field to match the namespace that Rook is
```
this can be explained only once we can move this to the top to keep common for both rbd and cephfs

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
````diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
+"rook-ceph".
+
+```console
+kubectl create -f cluster/examples/kubernetes/ceph/csi/rbd/snapshotclass.yaml
````
why we are creating rbd snapshotclass for cephfs?

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
````diff
@@ -99,3 +100,78 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
+```console
+kubectl delete -f cluster/examples/kubernetes/ceph/csi/cephfs/pvc-restore.yaml
+kubectl delete -f cluster/examples/kubernetes/ceph/csi/cephfs/snapshot.yaml
+kubectl delete -f cluster/examples/kubernetes/ceph/csi/cephfs/snapshotclass.yaml
````
please add the limitations for cephfs snapshots to this doc so that people will be aware of it
* Limit of 400 snapshots per filesystem
* PVC cannot be deleted if it's having snapshots

Madhu-1 commented on tests/framework/installer/ceph_csi_templates.go:
```diff
@@ -332,6 +332,21 @@ const (
+              args:
+                - "--csi-address=$(ADDRESS)"
+                - "--v=5"
+                - "--timeout=60s"
```
why the 60s? leader election flags are missing here

humblec replied:
Increased to 150. also added leader election for both cephfs and RBD as it was missing. :+1:

Madhu-1 commented on pkg/operator/ceph/csi/spec.go:
```diff
@@ -96,7 +96,7 @@ var (
 var (
 	// image names
-	DefaultCSIPluginImage   = "quay.io/cephcsi/cephcsi:v3.0.0"
+	DefaultCSIPluginImage   = "quay.io/cephcsi/cephcsi:v3.1.0"
```
need to add same to https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/version.go#L32

Madhu-1 replied:
same change need to be done for helm charts, operator.yaml openshift-operator.yaml and also need to updated the upgrade-documentation also

humblec commented:
@Madhu-1 comments are addressed. Flag based enablement can be done later or for next patch release of rook.

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
```diff
@@ -4,23 +4,19 @@ weight: 3250
-### SnapshotClass
-
-You need to create the `SnapshotClass`. The purpose of a `SnapshotClass` is
+- Create the `SnapshotClass`. The purpose of a `SnapshotClass` is
```
the snapshot class cannot be `Prerequisites` isn't it? as we are talking about how to create snapshotclass below.

humblec replied:
Adjusted this a bit.. Ptal.

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
```diff
@@ -99,3 +105,77 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
+
+## CephFS Snapshots
+
+NOTE: At present, there is a limit of 400 snapshots per cephFS filesystem. Also PVC cannot be deleted if it's having snapshots.
```
can we highlight this one?

humblec replied:
sure. done

Madhu-1 commented on cluster/charts/rook-ceph/values.yaml:
```diff
@@ -239,7 +239,7 @@ csi:
   #kubeletDirPath: /var/lib/kubelet
   #cephcsi:
-    #image: quay.io/cephcsi/cephcsi:v3.0.0
+    #image: quay.io/cephcsi/cephcsi:v3.1.0
```
the same change is also required in operator.yaml and openshift-operator.yaml

Madhu-1 commented on tests/framework/installer/ceph_csi_templates.go:
```diff
@@ -332,6 +333,22 @@ const (
+              args:
+                - "--csi-address=$(ADDRESS)"
+                - "--v=5"
+                - "--timeout=60s"
```
still, I see 60sec here

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
```diff
@@ -99,3 +101,82 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
+## CephFS Snapshots
+
+**NOTE: At present, there is a limit of 400 snapshots per cephFS filesystem. Also PVC cannot be deleted if it's having snapshots.
+make sure the snapshots on the PVC are deleted before you delete the PVC.**
```
```suggestion
Make sure all the snapshots on the PVC are deleted before you delete the PVC.**
```

humblec replied:
sure.. done

Madhu-1 commented on Documentation/ceph-csi-snapshot.md:
```diff
@@ -4,23 +4,19 @@ weight: 3250
-### SnapshotClass
-
-You need to create the `SnapshotClass`. The purpose of a `SnapshotClass` is
+- We also need a `SnapshotClass` for volume snapshot to work. The purpose of a `SnapshotClass` is
```
is this a prereq?

humblec replied:
Should be fine. As snapshot can not be created without SnapshotClass and also as we have more details in the coming sections.

humblec commented on Documentation/ceph-csi-snapshot.md:
```diff
@@ -99,3 +101,82 @@ kubectl delete -f cluster/examples/kubernetes/ceph/csi/rbd/pvc-restore.yaml
+the `csi.storage.k8s.io/snapshotter-secret-name` parameter should reference the
+name of the secret created for the cephfsplugin.
+
+In the snapshotclass, update the value of the `clusterID` field to match the namespace that Rook is
```
Sure.. we could do a follow-up PR for this.

humblec commented:
@travisn addressed the comments and also squashed 3/4 commits. PTAL.  The Ceph CSI v3.1.0 is also available now https://github.com/ceph/ceph-csi/releases/tag/v3.1.0 :+1:

Madhu-1 commented on cluster/charts/rook-ceph/templates/clusterrole.yaml:
```diff
@@ -296,6 +296,24 @@ rules:
   - apiGroups: [""]
     resources: ["persistentvolumeclaims/status"]
     verbs: ["update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
```
User cannot create and delete snapshots

Madhu-1 replied:
Yes snapshotter will keep writing error to logs
